src/components/primitives/mux-video.tsx

```diff
@@ -1,47 +1,86 @@
-import Hls, { Events } from 'hls.js'
+import type THls from 'hls.js'
 import * as React from 'react'
 import mergeRefs from 'react-merge-refs'
 
+import { createObserver } from '~/hooks/use-intersection-observer'
+
 export type MuxVideoProps = {
   muxSrc: string
+  lazy?: boolean
+  lazyObserverOptions?: IntersectionObserverInit
 } & Omit<JSX.IntrinsicElements['video'], 'src'>
 
+const supportsHls = (videoElm: HTMLVideoElement) =>
+  videoElm.canPlayType('application/vnd.apple.mpegurl')
+
 export const MuxVideo = React.forwardRef<HTMLVideoElement, MuxVideoProps>(
-  ({ muxSrc, className, ...rest }, ref) => {
+  ({ muxSrc, lazy = true, lazyObserverOptions, className, ...rest }, ref) => {
+    const hls = React.useRef<THls | null>(null)
     const videoRef = React.useRef<HTMLVideoElement>(null)
 
-    React.useEffect(() => {
-      let hls: Hls
+    const loadVideo = React.useCallback(
+      (videoElm: HTMLVideoElement) => {
+        if (hls.current) {
+          hls.current?.loadSource(muxSrc)
+        } else if (videoElm && supportsHls(videoElm)) {
+          videoElm.src = muxSrc
+        } else {
+          console.error('Unable to reproduce video')
+        }
+      },
+      [muxSrc]
+    )
 
-      if (videoRef.current) {
-        const video = videoRef.current
+    const importHls = React.useCallback(
```

Comment:
we should move this function inside `useEffect` as we're only using it there

Reply:
makes sense